src/Microsoft.AspNetCore.Routing/Internal/DataSourceDependantCache.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Routing.Internal
+{
+    internal class DataSourceDependantCache<T> where T : class
```

Comment:
I predict that we'll have plenty of cases where we want to use something like this...

Routing data can change, but rarely does, and it's a push model. We want to always be able to synchronously get the current data, but be able to reinitialize in the background when a change occurs.

Reply:
It's a new thing.

The route table and set of endpoints are fixed, it can be updated while the application is running. Razor pages is an example of this. You can 'drop in' a page to a running application and it will compile the page and add the appropriate entry to the route table.

Reply:
Something related to my comment earlier in this PR. Do we have scenarios for dynamic changing of endpoints apart from Razor pages or MVC? Should we do thing iteratively gather perf numbers along the way?

Reply:
Spoke to Kiran offline about this. I think that this is a critical requirement, and the design here isolates us well from the perf impact of this.